prog/vnet/update_load_balancer_node.rb

```diff
@@ -29,6 +41,70 @@ def before_run
     pop "load balancer is updated"
   end
 
+  label def put_certificate
+    cert = load_balancer.active_cert
+    cert_payload = cert.cert
+    cert_key_payload = OpenSSL::PKey::RSA.new(cert.csr_key).to_pem
+    vm.vm_host.sshable.cmd("sudo ip netns exec #{vm.inhost_name} mkdir -p #{cert_folder}")
+    vm.vm_host.sshable.cmd("sudo ip netns exec #{vm.inhost_name} tee #{cert_path}", stdin: cert_payload)
+    vm.vm_host.sshable.cmd("sudo ip netns exec #{vm.inhost_name} tee #{key_path}", stdin: cert_key_payload)
+    vm.vm_host.sshable.cmd("sudo ip netns exec #{vm.inhost_name} chown -R #{vm.inhost_name}:#{vm.inhost_name} #{cert_folder}")
+    hop_start_certificate_server
+  end
+
+  label def start_certificate_server
+    service_name = "#{vm.inhost_name}_cert_server"
+    service_file_path = "/vm/#{vm.inhost_name}/#{service_name}.service"
+    systemd_service = <<SYSTEMD
+[Unit]
+Description=Certificate Server
+After=network.target
+
+[Service]
+NetworkNamespacePath=/var/run/netns/#{vm.inhost_name}
+ExecStart=busybox httpd -f -p [FD00:0B1C:100D:CE::]:8080 -h #{cert_folder}
```

Comment:
interesting choice. One thing I'm wondering is:

1) do we really want a memory-unsafe HTTP server
2) can whatever http server used protect the system from very large or other kinds of malicious requests?

Reply:
1. I mainly chose this one because it was given by perplexity as the most lightweight web server. Maybe we can introduce the memory limit into the systemd file?
2. Considering the address we bind is completely blocked for non sudo user, the malicious actor would be the customer or a program they run in sudo. Still we need to rate-limit I guess. 

As an alternative, there is caddy from ZeroSSL https://github.com/caddyserver/caddy. It is written in GO. It also has a rate-limiting module but from a third party and it requires a re-build of caddy (https://github.com/RussellLuo/caddy-ext/tree/master/ratelimit) :/ 

I guess we can still use caddy (or if you have any other suggestion) + systemd configs.

Reply:
I found another way, we can still install caddy from the package but implement rate-limiting via nftables. This would give us static rate limiting. We can also configure caddy to limit the request size. How does it sound?

Reply:
I added caddy switch to this commit, it's still missing the rate-limiter, what do you think?
https://github.com/ubicloud/ubicloud/commit/f9daf854d9c32e510f54fcd3d37d630ee32e8e02

Reply:
Okay, probably an improvement, of my informal survey of web servers a year ago, caddy gave some decent vibes (and has a request size limit, we can set it to, like 1-8K presumably). But uh, I kinda feel like we should do what @byucesoy and I did for DNS servers to choose the implementation, that is, identify a shortlist of projects and analyze their development process and code size. In my recollection, most can be dismissed out of hand: caddy is a decent bar to set. Notably absent from my survey for no particular reason was the rust and crystal server approaches.

Reply:
I was doing a quick survey on the rust/crystal o'sphere and I'm not seeing anything super compelling, but I haven't done it systematically. I'd say 80% chance of caddy, but screw down the request size and take a look at the memory "RES" situation.

Reply:
Looking into this further and without properly testing myself, here is AI supported analysis;
1. Caddy, written in GO, kind of large code base, actively developed, feature rich, possibly memory usage on the high end.
2. Warp, written in Rust, small code base (25k loc), actively developed, not so feature rich, memory usage is low. The problem is, we'll need to compile and run this ourselves, binary is not available.
3. Kemal, written in Crystal, very small code base (5k), not much development activity, not so feature rich, memory usage is expected to be higher than warp but lower than caddy. The problem is, we'll need to compile and run this ourselves, binary is not available.

There are some alternatives that are written in C but we avoid them. I also found some written in python but I assume we avoid that, too.

Giving it some more thought, I'm thinking, we already have ruby environment setup, we are already experienced in puma, it may require more memory consumption but, what about it? We can also consider puma for the sake of developer familiarity. 

I need to perform controlled tests to verify each of these memory consumptions.

Reply:
Alright, here are my findings
1. busybox;
```
[root@vhgwk6zhtcqj4jh8f1we0wet1m vmv4xy42]# pidstat --human -r -u -p 993481 10
Linux 5.15.0-100-generic (vhgwk6zhtcqj4jh8f1we0wet1m) 	08/21/2024 	_x86_64_	(12 CPU)
^[
02:42:46 PM   UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
02:42:56 PM 33540    993481    0.0%    0.0%    0.0%    0.0%    0.0%    10  busybox

02:42:46 PM   UID       PID  minflt/s  majflt/s     VSZ     RSS   %MEM  Command
02:42:56 PM 33540    993481      0.00      0.00    2.4M    4.0k   0.0%  busybox

02:42:56 PM   UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
02:43:06 PM 33540    993481    0.0%    0.0%    0.0%    0.0%    0.0%    10  busybox

02:42:56 PM   UID       PID  minflt/s  majflt/s     VSZ     RSS   %MEM  Command
02:43:06 PM 33540    993481      0.00      0.00    2.4M    4.0k   0.0%  busybox

02:43:06 PM   UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
02:43:16 PM 33540    993481    0.1%    0.0%    0.0%    0.0%    0.1%    10  busybox
```
2. Warp;
```
^C[root@vhgwk6zhtcqj4jh8f1we0wet1m vmv4xy42]# pidstat --human -r -u -p 1005520 10
Linux 5.15.0-100-generic (vhgwk6zhtcqj4jh8f1we0wet1m) 	08/21/2024 	_x86_64_	(12 CPU)

04:01:57 PM   UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
04:02:07 PM     0   1005520    0.1%    0.0%    0.0%    0.0%    0.1%     3  warp_server

04:01:57 PM   UID       PID  minflt/s  majflt/s     VSZ     RSS   %MEM  Command
04:02:07 PM     0   1005520      0.50      0.00  870.0M    6.2M   0.0%  warp_server

04:02:07 PM   UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
04:02:17 PM     0   1005520    0.0%    0.0%    0.0%    0.0%    0.0%     3  warp_server

04:02:07 PM   UID       PID  minflt/s  majflt/s     VSZ     RSS   %MEM  Command
04:02:17 PM     0   1005520      0.10      0.00  870.0M    6.2M   0.0%  warp_server
```

3. Kemal
```
[root@vhgwk6zhtcqj4jh8f1we0wet1m ~]# pidstat --human -r -u -p 992684 10
Linux 5.15.0-100-generic (vhgwk6zhtcqj4jh8f1we0wet1m) 	08/21/2024 	_x86_64_	(12 CPU)

02:38:47 PM   UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
02:38:57 PM     0    992684    0.0%    0.0%    0.0%    0.0%    0.0%     8  crystal

02:38:47 PM   UID       PID  minflt/s  majflt/s     VSZ     RSS   %MEM  Command
02:38:57 PM     0    992684      0.00      0.00  647.7M  547.1M   0.4%  crystal

02:38:57 PM   UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
02:39:07 PM     0    992684    0.0%    0.0%    0.0%    0.0%    0.0%     8  crystal

02:38:57 PM   UID       PID  minflt/s  majflt/s     VSZ     RSS   %MEM  Command
02:39:07 PM     0    992684      0.00      0.00  647.7M  547.1M   0.4%  crystal

02:39:07 PM   UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
02:39:17 PM     0    992684    0.0%    0.0%    0.0%    0.0%    0.0%     8  crystal

02:39:07 PM   UID       PID  minflt/s  majflt/s     VSZ     RSS   %MEM  Command
02:39:17 PM     0    992684      0.00      0.00  647.7M  547.1M   0.4%  crystal
```

4. Caddy
```
[root@vhgwk6zhtcqj4jh8f1we0wet1m vmv4xy42]# pidstat --human -r -u -p 437776 10
Linux 5.15.0-100-generic (vhgwk6zhtcqj4jh8f1we0wet1m) 	08/21/2024 	_x86_64_	(12 CPU)

04:04:58 PM   UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
04:05:08 PM   998    437776    0.0%    0.0%    0.0%    0.0%    0.0%    10  caddy

04:04:58 PM   UID       PID  minflt/s  majflt/s     VSZ     RSS   %MEM  Command
04:05:08 PM   998    437776      0.00      0.00    1.2G   17.5M   0.0%  caddy

04:05:08 PM   UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
04:05:18 PM   998    437776    0.0%    0.0%    0.0%    0.0%    0.0%    10  caddy

04:05:08 PM   UID       PID  minflt/s  majflt/s     VSZ     RSS   %MEM  Command
04:05:18 PM   998    437776      0.00      0.00    1.2G   17.5M   0.0%  caddy
```

So, according to these results, if we only consider RSS, the order is;
1. busybox (4k)
2. Warp (6.2M)
3. Caddy (17.5M)
4. Kemal (547.1M)

if we consider VSZ;
1. busybox (2.4M)
2. Kemal (647.7M)
3. Warp (870M)
4. Caddy (1.2G)
Since busybox is not memory safe, I feel like we are picking either Warp or Caddy. Considering the activity, being backed by ZeroSSL itself, I am more inclined to go with Caddy but 1.2G Virtual memory is huge! Plus, the codebase is very big. So, I'm happy to go with warp, too. 

Another alternative, I feel like we can limit memory size via systemd and use busybox, too.

Reply:
I'm not sure if it can easily serve static files, but @bsatzger is using [cloudflare/pingora](https://github.com/cloudflare/pingora) for API gateway implementation. We could consider introducing a single dependency

Reply:
We're not doing a proxy thing though, right? It's serving a small piece of text. And maybe other metadata in the future.

Here's a dumb thought: how about Go's standard library web server?

https://www.perplexity.ai/search/write-me-a-go-program-that-loa-Hi9Q69k1QoO7ARp8ZPeOwQ

```go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"time"

	"golang.org/x/time/rate"
)

// Define the rate limiter
var limiter = rate.NewLimiter(10, 1) // 10 requests per second

func certHandler(w http.ResponseWriter, r *http.Request) {
	// Check if the request can proceed
	if !limiter.Allow() {
		http.Error(w, "Too Many Requests", http.StatusTooManyRequests)
		return
	}

	// Limit the size of the request body
	r.Body = http.MaxBytesReader(w, r.Body, 1024) // Limit to 1KB

	// Read the cert.pem file
	data, err := ioutil.ReadFile("cert.pem")
	if err != nil {
		http.Error(w, "File not found", http.StatusNotFound)
		return
	}

	// Write the file contents to the response
	w.Header().Set("Content-Type", "application/x-pem-file")
	w.WriteHeader(http.StatusOK)
	w.Write(data)
}

func main() {
	// Set up the HTTP server
	http.HandleFunc("/cert.pem", certHandler)

	// Start the server
	port := ":8080"
	fmt.Printf("Starting server on port %s\n", port)
	if err := http.ListenAndServe(port, nil); err != nil {
		log.Fatalf("Could not start server: %s\n", err)
	}
}
```

Reply:
But uh, basically, I had these general goal for the web server: rate limiting, small code, institutional stability, memory safety, maybe some graceful programming option.

Caddy has rate limiting, memory safety, and institutional stability.

I think the Rust options still kind of froth around, so without radically smaller sizes (much closer to busybox) it's harder to get excited about those.

But for an endpoint or five, perhaps straight up Go net/http will be a good set-and-forget option, even more so than a much larger and more complex program in Caddy.

Reply:
I had Perplexity give some ideas on writing tests https://www.perplexity.ai/search/write-me-a-go-program-that-loa-Hi9Q69k1QoO7ARp8ZPeOwQ#2

Reply:
I have done that. the new server lives in https://github.com/ubicloud/cert-server. We consume it here through github packages once and then copy the binary locally per VM.

Reply:
should it be cert-server or should this be the metadata-endpoint server (or some other name less customary but also less...insipid?)

For example, finally, at some length, we could use `cloud-init` in a non-basic capacity. (does it matter? should we use `NoCloud` forever?!) 

I realize all we need is two HTTP endpoints today, but I suspect we're going to want to surface more stuff (should it be like `load-balancer/cert.pem`?)

I think it's very good that the simple case does not involve any dynamic text, like "load-balancer/#{lbidhere}`" because it's great to be able to hard code this. People with advanced use cases and multiple load balancers can deal with that in a more elaborate way.

All in all, these comments mean only two things:

1. Should we add an additional constant string to the URI to contain load-balancer related metadata data?
2. Maybe the project could be named differently, but that's not an interface with the user, so, kind of whatever, it's up to you if you feel like changing binary names later or whatnot.

Reply:
I did what I could to make a new name, but none of the alternates I could get an LLM to generate really offer much in the way of improvement. Interesting quick overview of the various alternatives, though:

https://www.perplexity.ai/search/tell-me-about-metadata-endpoin-3ssyS24iRbWUqDUFB8XiNw